# GEL NMD analysis
This repostitory describes the workflow to identify enrichment and recurrence of _de novo_ variants in genes with regional nonsense constraint.

## Citation
Please cite our preprint:

Blakes, A. J. M., Whiffin, N., Johnson, C. A., Ellingford, J. & Banka, S. [Regional nonsense constraint offers clinical and biological insights into rare genetic disorders](https://doi.org/10.1101/2024.10.10.24315185). 2024.10.10.24315185 (2024).

## Code and data availability
Our analysis of *de novo* variation was performed in trio sequencing data from the 100,000 Genomes Project (100KGP), the UK Genomic Medicine Service (GMS), and the [Deciphering Developmental Disorders](https://www.ddduk.org/) study.

Data from the 100KGP and the GMS was accessed through the [National Genomic Research Library](https://www.genomicsengland.co.uk/blog/genomics-101-what-is-the-national-genomic-research-library). Access to the NGRL is available to registered academics with approved projects, as described [here](https://www.genomicsengland.co.uk/join-us).

## Workflow export
The most up to data version of this analysis is currently awaiting export approval form the Genomics England secure research environment. This repo will be updated upon approval being granted. Last checked 11/12/25.

The newer version of this repository is described by a Make workflow. This README will be updated to describe this workflow once the export has been approved. 

### Update 11/12/2025
Export of the analysis workflow from the Genomics England secure research environment is being prevented by an internal error in the Genomics England export service. Genomics England are working to rectify this issue.

## Companion repository
See also the [main repository here](https://github.com/alexblakes/regional_nonsense_constraint).
